### PR TITLE
feature: read maxon epos' special Sdos

### PR DESCRIPTION
**Pull request by zoenglinghou**
Added functions to read `bool` and the diagnostic message `array`.

I might have to revert some auto formatting tho, if there is any `.clang-format` to stick to.

### CODE REVIEW

**Comment by jungerj** [comment]
There should not be any drive-specific implementation details here (i.e. no ...maxon...).
At second thought I also don't like the bool SDO read. The size of bool is actually [implementation specific](https://en.cppreference.com/w/cpp/language/types). You should send a uint of whatever size the hardware expects for a boolean type (probably 1 byte) and check if the result is zero or not.

I think we got into a bit of a template hell. At the moment we first have the type-agnostic sendSdoRead / sendSdoWrite in the BusBase (implemented in the header), then a templated call to it in the SlaveBase (*NOT* implemented in the header and never used), then we have the type-suffixed SDO reads / writes in the SlaveBase calling the type agnostic function in BusBase and a templated sendSdoRead / sendSdoWrite in the EthercatDevice, implemented again in the source file with all type specifications. In the end everything is however passed as void* to SOEM, without any checks at all. I think we should clear these issues before merging.

I will implement two possible versions to remove the template hell and discuss them with @firesurfer.

**--> We will most certainly not need the two new SDO calls in the future.**

File: soem_interface/src/soem_interface/EthercatSlaveBase.cpp [hunk 1]
```diff
@@ -24,84 +24,101 @@
...
-                                                        const uint32_t value);
-template bool EthercatSlaveBase::sendSdoWrite<uint64_t>(const uint16_t index, const uint8_t subindex, const bool completeAccess,
-                                                        const uint64_t value);
-template bool EthercatSlaveBase::sendSdoWrite<float>(const uint16_t index, const uint8_t subindex, const bool completeAccess,
-                                                     const float value);
-template bool EthercatSlaveBase::sendSdoWrite<double>(const uint16_t index, const uint8_t subindex, const bool completeAccess,
-                                                      const double value);
-template bool EthercatSlaveBase::sendSdoWrite<std::string>(const uint16_t index, const uint8_t subindex, const bool completeAccess,
-                                                           const std::string value);
+template bool EthercatSlaveBase::sendSdoWrite<int8_t>(const uint16_t index, const uint8_t subindex,
```
Comment (firesurfer) [inline soem_interface/src/soem_interface/EthercatSlaveBase.cpp:55]: Make sure that you dont just commit formatting changes.

**Comment by firesurfer** [comment]
@jungerj I agree with you.
Regarding supported types I would suggest you take a look at what the ethercat standard defines as types and implement those.
@zoenglinghou Perhaps see if you can replace these two sdo calls with an existing one?

**Comment by zoenglinghou** [comment]
@firesurfer I would see what I can do. If I can replace the functions, then there is **no** need for the PR and the formatting shenanigans...

btw is there a `.clang-format` out there? Auto formatting is just an editor thing and I wish I could have some sort of standard to follow

**Comment by tomlankhorst** [comment]
@zoenglinghou looks like Google style `clang-format -style=google -dump-config > .clang-format`

**Comment by jungerj** [comment]
@zoenglinghou Can you check whether you can build your latest version of the maxon sdk with [this](https://github.com/leggedrobotics/soem_interface/pull/4) version of the soem_interface and [this](https://github.com/leggedrobotics/ethercat_sdk_master/pull/4) version of the ethercat_sdk_master? The SDO templates there are in the header files, making your array and bool additions unnecessary.

**Comment by zoenglinghou** [comment]
@jungerj they built with no problem! Thanks for the good work!!!

**Comment by firesurfer** [comment]
@jungerj 
Can you please decide on this PR? As far as I can see @zoenglinghou didnt remove the maxon specific reads. 
I still completely agree that there shouldnt be any drive specific functions in this repo

**Comment by zoenglinghou** [comment]
@firesurfer the changes are no longer needed since @jungerj fixed it in #4 . I would close this PR.

**Pull request closed without merging**